### ISSUE TITLE
Proposal .asInterface() method for all go2hx types

### ISSUE THREAD

**Issue by elliott5** [go2hx/go2hx#133]
Would it be possible for all go2hx Go named types (be they implemented as classes or typedefs) to have a method called ".asInterface()" that delivers back the interface form of that type?

```haxe
      // inside the users normal Haxe code
      stdgo.fmt.Fmt.printf("%+v\n", someGo2hxThing.asInterface())
```

Because at the moment you would have to find out the type of someGo2hxThing before you can make it an interface:
```haxe
      // inside the users normal Haxe code
      stdgo.fmt.Fmt.printf("%+v\n", TheTypeOf_someGo2hxThing_asInterface.new(someGo2hxThing))
```
...which would be very irritating and reduce the speed at which you can code.

**Comment by PXshadow** [comment]
I agree, this would be very useful, an alternative to this would be a macro system such as:
```haxe
Go.asInterface(someGo2hxThing);
```

I do think I prefer your way in the proposal, I have to also check to see if it doesn't run into reference cycles. @elliott5

**Comment by elliott5** [comment]
Since `Go.asInterface(something)` would apply for all types, that would be preferable.

It would also mean that the `Someclass_asInterface` could be removed from the documentation. ;)

**Comment by PXshadow** [comment]
Trying to make it so interfaces can be made firstly without having to define all of the fields, then I can work on the macro to make it even easier.

**Comment by PXshadow** [comment]
Ran into some issues, types with no type params can use this method but ones with, will have to use the old method. The reason is because the macro needs to know the type params and so cannot be created before inside a funtion as then a generic type is only known, causing the static extension macro to fail creating a typedef for each typeparam and instead creates a recursive type of itself.

**Comment by elliott5** [comment]
That's a disappointment @PXshadow , but unsurprising given past experience.

For a generic type, could the code for `.asInterface()` be generated by the compiler when a generic type is instantiated:
```haxe
// haxe syntax probably wrong
var x = new(myType<typeParam>); 
x.asInterface = function() { /* code to implement specific to this typeParam */ };
```

**Comment by PXshadow** [comment]
> That's a disappointment @PXshadow , but unsurprising given past experience.
> 
> For a generic type, could the code for `.asInterface()` be generated by the compiler when a generic type is instantiated:
> ```haxe
> // haxe syntax probably wrong
> var x = new(myType<typeParam>); 
> x.asInterface = function():AnyInterface { /* code to implement specific to this typeParam */ };
> ```

@elliott5 great idea working on it. It's a much cleaner approach just need it to pass the regression issues.

**Comment by PXshadow** [comment]
```haxe
stdgo.Go.anyInterface
``

Is a very mature macro function nowadays and can be used in most cases quite reliably, including Haxe code needing to turn into ``stdgo.AnyInterface`` = any/interface{}

This is what a compile of a named struct{} looks like.
```go
type t struct{}
```
```haxe
package _internal.github_dot_com.go2hx.go4hx.rnd;
@:structInit class T_t {
    public function new() {}
    public function __underlying__() return stdgo.Go.toInterface(this);
    public function __copy__() {
        return new T_t();
    }
}
```

This will obviously not work for generic types acessing an unknown type, my idea to deal with this is to use a [@:genericBuild](https://haxe.org/manual/macro-generic-build.html) macro, to create new types per new typeParam.

**Comment by PXshadow** [comment]
``stdgo.Go.asInterface``
